src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1120 @@
+
+          const prototypeInputs: sb3.Block["inputs"] = {};
+          for (const arg of args) {
+            const shadowId = generateId();
+            blockData[shadowId] = {
+              opcode: prop(
+                {
+                  boolean: OpCode.argument_reporter_boolean,
+                  numberOrString: OpCode.argument_reporter_string_number
+                },
+                arg.type
+              ),
+
+              next: null,
+              parent: prototypeId,
+
+              inputs: {},
+              fields: {
+                VALUE: [arg.name]
+              },
+
+              shadow: true,
+              topLevel: false
+            };
+
+            prototypeInputs[arg.id] = [BIS.INPUT_SAME_BLOCK_SHADOW, shadowId];
+          }
+
+          blockData[prototypeId] = {
+            opcode: OpCode.procedures_prototype,
+
+            next: null,
+            parent: block.id,
+
+            inputs: prototypeInputs,
+            fields: {},
+
+            shadow: true,
+            topLevel: false,
+
+            mutation: {
+              tagName: "mutation",
+              children: [],
+              proccode: block.inputs.PROCCODE.value,
+              argumentids: JSON.stringify(args.map(arg => arg.id)),
+              argumentnames: JSON.stringify(args.map(arg => arg.name)),
+              argumentdefaults: JSON.stringify(args.map(arg => arg.default)),
+              warp: JSON.stringify(warp) as "true" | "false"
+            }
+          };
+
+          inputs.custom_block = [BIS.INPUT_SAME_BLOCK_SHADOW, prototypeId];
+
+          break;
+        }
+
+        case OpCode.procedures_call: {
+          const proccode = block.inputs.PROCCODE.value;
+          const customBlockData = customBlockDataMap[proccode];
+          if (!customBlockData) {
+            warn(
+              `Missing custom block prototype for proccode ${proccode} (${block.id} in ${target.name}); skipping this block`
+            );
+            return null;
+          }
+
+          const { args, warp } = customBlockData;
+
+          mutation = {
+            tagName: "mutation",
+            children: [],
+            proccode: block.inputs.PROCCODE.value,
+            argumentids: JSON.stringify(args.map(arg => arg.id)),
+            warp: JSON.stringify(warp) as "true" | "false"
+          };
+
+          const inputEntries = {};
+          const constructedInputs = {};
+          const initialValues = {};
+          for (let i = 0; i < args.length; i++) {
+            const { type, id } = args[i];
+            inputEntries[id] = prop(
+              {
+                boolean: sb3.BooleanOrSubstackInputStatus,
+                numberOrString: BIS.TEXT_PRIMITIVE
+              },
+              type
+            );
+            initialValues[id] = prop(
+              {
+                boolean: false,
+                numberOrString: ""
+              },
+              type
+            );
```

Comment:
```suggestion
            switch (type) {
              case "boolean":
                inputEntries[id] = sb3.BooleanOrSubstackInputStatus;
                initialValues[id] = false;
                break;
              case "numberOrString":
                inputEntries[id] = BIS.TEXT_PRIMITIVE;
                initialValues[id] = "";
                break;
            }
```

I *think* this does the same thing. Not sure if `args[i].type` is guaranteed to be `"boolean" | "numberOrString"`; if it's not then add a `default` which throws.

Reply:
Also, here the initial value is `false` (a boolean) but [here](https://github.com/PullJosh/sb-edit/pull/33/files#diff-9d556895c6d9cbaaf5813685827b42acR849) it's `"false"` (a string). Is this intentional?

Reply:
re: second comment - That's a good point. The initial value for a boolean...actually doesn't matter, because when empty boolean inputs are serialized, the passed value isn't considered at all. As such, it's technically not necessary to provide an initial value here at all. I haven't tested (I'll get back to messing with code tomorrow), but simply not setting `initialValues[id]` should be valid code.

As for whether it's necessary in argumentids (which is serialized directly from the `CustomBlockArg` data) - I'm not sure. I provide `"false"` because that's the value that Scratch 3.0 outputs. I would suggest it might be possible to simply output undefined, ala `["",,""]` in custom blocks, but JSON doesn't support output like that, so it's basically a moot point. Better to match the output of 3.0 in the same situation.